controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+		uiImage:        objectStoreUIImage,
+	}
+
+	ds.ObjectStoreInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueObjectStore,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueObjectStore(cur) },
+			DeleteFunc: osc.enqueueObjectStore,
+		},
+	)
+	osc.cacheSyncs = append(osc.cacheSyncs, ds.ObjectStoreInformer.HasSynced)
+
+	return osc
+}
+
+func (osc *ObjectStoreController) Run(workers int, stopCh <-chan struct{}) {
+	osc.logger.Info("Starting Longhorn Object Store Controller")
+	defer osc.logger.Info("Shut down Longhorn Object Store Controller")
+	defer osc.queue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync("longhorn object stores", stopCh, osc.cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(osc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (osc *ObjectStoreController) worker() {
+	for osc.processNextWorkItem() {
+	}
+}
+
+func (osc *ObjectStoreController) processNextWorkItem() bool {
+	key, quit := osc.queue.Get()
+	if quit {
+		return false
+	}
+	defer osc.queue.Done(key)
+
+	err := osc.syncObjectStore(key.(string))
+	if err == nil {
+		osc.queue.Forget(key)
+	} else if osc.queue.NumRequeues(key) < maxRetries {
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", retrying", err)
+		osc.queue.AddRateLimited(key)
+	} else {
+		utilruntime.HandleError(err)
+		osc.logger.WithError(err).Errorf("error syncing object store: \"%v\", giving up", err)
+		osc.queue.Forget(key)
+	}
+
+	return true
+}
+
+func (osc *ObjectStoreController) enqueueObjectStore(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Failed to get key for %v: %v", obj, err))
+		return
+	}
+	osc.queue.Add(key)
+}
+
+func (osc *ObjectStoreController) syncObjectStore(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	store, err := osc.ds.GetObjectStore(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil // already deleted, nothing to do
+		}
+		return err
+	}
+
+	if !store.DeletionTimestamp.IsZero() && store.Status.State != longhorn.ObjectStoreStateTerminating {
+		_, err = osc.setObjectStoreState(store, longhorn.ObjectStoreStateTerminating)
+		return err
+	}
+
+	switch store.Status.State {
+	case longhorn.ObjectStoreStateStarting, longhorn.ObjectStoreStateError:
+		return osc.handleStarting(store)
+
+	case longhorn.ObjectStoreStateRunning:
+		return osc.handleRunning(store)
+
+	case longhorn.ObjectStoreStateStopping:
+		return osc.handleStopping(store)
+
+	case longhorn.ObjectStoreStateStopped:
+		return osc.handleStopped(store)
+
+	case longhorn.ObjectStoreStateTerminating:
+		return osc.handleTerminating(store)
+
+	default:
+		return osc.initializeObjectStore(store)
+	}
+}
+
+// This function handles the case when the object store is in "Starting"
+// state. That means, all resources have been created in the K8s API, but the
```

Comment:
> That means, all resources have been created in the K8s API

This comment is a bit misleading. The dependent resources might not have been created yet and we will create them in this function